src/main/java/com/google/sps/servlets/Caption.java

```diff
@@ -0,0 +1,53 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.io.InputStream;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+public class Caption {
+  /*
+   * Usage: String caption = new Caption().getCaptionFromID("abcdefg");
+   * 
+   * Return null if not found.
+   */
+
+  public CaptionService captionService;
+
+  public Caption() {
+    captionService = new CaptionService();
+  }
+
+  public void setCaptionService(CaptionService captionService) {
+    this.captionService = captionService;
+  }
+
+  public String getCaptionFromId(String videoId) {
+    // BCP47 language tags
+    final String[] languages = {"en", "en-US", "en-GB", "en-AU", "en-IE", "en-ZA"};
+    for (String language : languages) {
+      try {
+        return parseXmlFromStream(captionService.fetchStream(videoId, language));
+      } catch (SAXException | ParserConfigurationException | IOException
+          | IllegalArgumentException e) {
+        continue;
+      }
```

Comment:
Since you are explicitly catching these specific exceptions, then you need to test each one of them explicitly. You would need to mock parseXmlFromStream or any of the methods inside that method that actually throw those errors to actually throw them and then test getCaptionFromId returns null. I see you doing this already for IOException. So keep it up with the others.

Again, since you an explicit list of exceptions to catch, then maybe you should throw a different one in a test and verify that getCaptionFromId eventually throws a different one?

Reply:
SAXException, ParseConfigurationException, and IllegalArgumentException are thrown by the private function `parseXmlFromStream` and handled(ignored) in the public function `getCaptionFromId`. I am not sure if I should test a private function and mock its behavior of throwing exceptions?

Reply:
Yup you want to test that your code ignores these exceptions. Is it possible to mock the methods thats throw these exceptions  and throw them rather than mocking the private method?

Reply:
Good idea. I mocked the DocumentBuilderFactory that throws the execeptions and it worked!